firmware/isp.c

```diff
@@ -124,21 +125,31 @@ void ispConnect() {
 
 void ispDisconnect() {
 
+	cli();
 	/* set all ISP pins inputs */
 	ISP_DDR &= ~((1 << ISP_RST) | (1 << ISP_SCK) | (1 << ISP_MOSI));
 	/* switch pullups off */
 	ISP_OUT &= ~((1 << ISP_RST) | (1 << ISP_SCK) | (1 << ISP_MOSI));
+	sei();
 
 	/* disable hardware SPI */
 	spiHWdisable();
 }
 
 void ispSafeResetPulse(void)
 {
+	/* 
```

Comment:
In light of what you explained, I think it might be better to clear/set each bit individually, e.g. the code in ispDisconnect could look like this:
       /* set all ISP pins inputs */
        ISP_DDR &= ~(1 << ISP_RST);
        ISP_DDR &= ~(1 << ISP_SCK);
        ISP_DDR &= ~(1 << ISP_MOSI);
        /* switch pullups off */
        ISP_OUT &= ~(1 << ISP_RST);
        ISP_OUT &= ~(1 << ISP_SCK);
        ISP_OUT &= ~(1 << ISP_MOSI);
I think it is correct too, since I see no updates that require the hardware port bits to be updated simultaneously (to avoid jitter). It would give better interrupt latency as there is no cli/sei. Don't know how important this is as routines like ispDisconnect() are called outside the heavy usb traffic (reading and writing flash). Nevertheless the software usb implementation is full of things that e.g. require an exact nr of instructions, so it might be safer... It is also a few  instructions less...